mvc/src/main/java/nextstep/mvc/controller/tobe/AnnotationHandlerMapping.java

```diff
@@ -1,30 +1,72 @@
 package nextstep.mvc.controller.tobe;
 
 import jakarta.servlet.http.HttpServletRequest;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Set;
 import nextstep.mvc.HandlerMapping;
+import nextstep.web.annotation.Controller;
+import nextstep.web.annotation.RequestMapping;
+import nextstep.web.support.RequestMethod;
+import nextstep.web.support.RequestUrl;
+import org.reflections.Reflections;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.HashMap;
-import java.util.Map;
-
 public class AnnotationHandlerMapping implements HandlerMapping {
 
     private static final Logger log = LoggerFactory.getLogger(AnnotationHandlerMapping.class);
 
     private final Object[] basePackage;
-    private final Map<HandlerKey, HandlerExecution> handlerExecutions;
+    private final HandlerExecutions handlerExecutions;
 
     public AnnotationHandlerMapping(final Object... basePackage) {
         this.basePackage = basePackage;
-        this.handlerExecutions = new HashMap<>();
+        this.handlerExecutions = new HandlerExecutions(new HashMap<>());
     }
 
     public void initialize() {
+        Reflections reflections = new Reflections(basePackage);
+        Set<Class<?>> annotatedControllers = reflections.getTypesAnnotatedWith(Controller.class);
+        for (Class<?> controller : annotatedControllers) {
+            putControllerToHandlerExecutions(controller);
+        }
+
         log.info("Initialized AnnotationHandlerMapping!");
     }
 
     public Object getHandler(final HttpServletRequest request) {
-        return null;
+        RequestUrl requestURI = new RequestUrl(request.getRequestURI());
+        RequestMethod requestMethod = RequestMethod.from(request.getMethod());
+        return handlerExecutions.get(requestURI, requestMethod);
+    }
+
+    private void putControllerToHandlerExecutions(Class<?> controller) {
+        Method[] controllerMethods = controller.getDeclaredMethods();
+        Object instance = getInstance(controller);
+        for (Method method : controllerMethods) {
+            RequestMapping requestMapping = method.getDeclaredAnnotation(RequestMapping.class);
+            putHandlerKeyAndExecution(instance, method, requestMapping);
+        }
+    }
+
+    private Object getInstance(Class<?> controller) {
+        Object o;
+        try {
+            o = controller.getConstructor().newInstance();
+        } catch (InstantiationException | IllegalAccessException | InvocationTargetException |
+                 NoSuchMethodException e) {
+            throw new RuntimeException(e);
+        }
+        return o;
+    }
```

Comment:
```suggestion
    private Object getInstance(Class<?> controller) {
        try {
            return controller.getConstructor().newInstance();
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException |
                 NoSuchMethodException e) {
            throw new RuntimeException(e);
        }
    }
```
R: try에서 바로 return할 수도 있을 것 같은데, 로컬 변수로 선언해 return한 이유가 궁금해요~

추가적으로 checked exception을 RuntimeException으로 변환해 처리하신 이유도 궁금합니다~

Reply:
이유라기보다는 해당 메서드가 나온 정황을 말씀드리자면 IDE 의 도움을 받아 `.var` 로 변수를 만들어 주었고 IDE 의 도움을 또 받아 try catch 를 자동으로 만들어 주는 과정에서 발생한 코드입니다,, 잘못을 알았으니 앞으로 주의하겠습니다.

RuntimeException 으로 generic 하게 한 이유는 catch 되는 exception 이 많아서 그랬는데, 조금 더 명시적으로 예외를 처리하기 위해 InstanceInitializationException 이라는 클래스를 만들어 예외를 처리하였습니다!

지적 감사드립니다 🙇